File: Runtime/TransitionAdjacencyMultiplierMap.vb
'************************************************************************************
' ST-Sim: A .NET class library for developing state-and-transition simulation models.
'
' Copyright © 2007-2017 Apex Resource Management Solution Ltd. (ApexRMS). All rights reserved.
'
'************************************************************************************

Imports SyncroSim.Core
Imports SyncroSim.StochasticTime

Class TransitionAdjacencyMultiplierMapEntry
    Public Map As New Dictionary(Of Double, TransitionAdjacencyMultiplier)
    Public Items As New SortedList(Of Double, TransitionAdjacencyMultiplier)
End Class

Class TransitionAdjacencyMultiplierMap
    Inherits STSimMapBase3(Of TransitionAdjacencyMultiplierMapEntry)

    Private m_DistributionProvider As STSimDistributionProvider

    Public Sub New(
        ByVal scenario As Scenario,
        ByVal multipliers As TransitionAdjacencyMultiplierCollection,
        ByVal distributionProvider As STSimDistributionProvider)

        MyBase.New(scenario)

        Me.m_DistributionProvider = DistributionProvider

        For Each Item As TransitionAdjacencyMultiplier In multipliers
            Me.AddMultiplier(Item)
        Next

    End Sub

    Public Function GetAdjacencyMultiplier(
        ByVal transitionGroupId As Integer,
        ByVal stratumId As Integer,
        ByVal secondaryStratumId As Nullable(Of Integer),
        ByVal iteration As Integer,
        ByVal timestep As Integer,
        ByVal attributeValue As Double) As Double

        Dim e As TransitionAdjacencyMultiplierMapEntry =
            Me.GetItem(
                transitionGroupId,
                stratumId,
                secondaryStratumId,
                iteration,
                timestep)

        If (e Is Nothing) Then
            Return 1.0
        End If

        Debug.Assert(e.Items.Count = e.Map.Count)

        If (e.Map.Count = 1) Then

            Dim tam As TransitionAdjacencyMultiplier = e.Items.First.Value
            tam.Sample(iteration, timestep, Me.m_DistributionProvider, DistributionFrequency.Always)

            Return tam.CurrentValue.Value

        End If

        If (e.Map.ContainsKey(attributeValue)) Then

            Dim tam As TransitionAdjacencyMultiplier = e.Map(attributeValue)
            tam.Sample(iteration, timestep, Me.m_DistributionProvider, DistributionFrequency.Always)

            Return tam.CurrentValue.Value

        End If

        Dim PrevVal As Double = -1.0
        Dim NextVal As Double = -1.0

        Debug.Assert(attributeValue >= 0.0)

        For Each k As Double In e.Items.Keys

            Debug.Assert(k <> attributeValue)

            If (k > attributeValue) Then
                NextVal = k
                Exit For
            End If

            PrevVal = k

        Next

        If (NextVal = -1.0) Then
            NextVal = PrevVal
        End If

        Dim PrevMult As TransitionAdjacencyMultiplier = e.Map(PrevVal)
        Dim NextMult As TransitionAdjacencyMultiplier = e.Map(NextVal)

        PrevMult.Sample(iteration, timestep, Me.m_DistributionProvider, DistributionFrequency.Always)
        NextMult.Sample(iteration, timestep, Me.m_DistributionProvider, DistributionFrequency.Always)

        If (PrevVal <> -1.0) Then

            If NextVal <> -1.0 Then

                If (PrevVal = NextVal) Then
                    Return PrevMult.CurrentValue.Value
                Else

                    Return Interpolate(
                        PrevVal,
                        PrevMult.CurrentValue.Value,
                        NextVal,
                        NextMult.CurrentValue.Value,
                        attributeValue)

                End If

            Else
                Return PrevMult.CurrentValue.Value
            End If

        Else
            Return 1.0
        End If

    End Function

    Private Sub AddMultiplier(ByVal item As TransitionAdjacencyMultiplier)

        Dim e As TransitionAdjacencyMultiplierMapEntry =
            Me.GetItemExact(
                item.TransitionGroupId,
                item.StratumId,
                item.SecondaryStratumId,
                item.Iteration,
                item.Timestep)

        If (e Is Nothing) Then

            e = New TransitionAdjacencyMultiplierMapEntry()

            Me.AddItem(
                item.TransitionGroupId,
                item.StratumId,
                item.SecondaryStratumId,
                item.Iteration,
                item.Timestep,
                e)

        End If

        If (Not e.Map.ContainsKey(item.AttributeValue)) Then

            Debug.Assert(Not e.Items.ContainsKey(item.AttributeValue))

            e.Map.Add(item.AttributeValue, item)
            e.Items.Add(item.AttributeValue, item)

        End If

        Debug.Assert(Me.HasItems())

    End Sub

End Class

File: Runtime/TransitionAdjacencyMultiplierMap.cs
// A SyncroSim Package for developing state-and-transition simulation models using ST-Sim.
// Copyright © 2007-2018 Apex Resource Management Solution Ltd. (ApexRMS). All rights reserved.

using System.Linq;
using SyncroSim.Core;
using System.Diagnostics;
using SyncroSim.StochasticTime;
using System.Collections.Generic;

namespace SyncroSim.STSim
{
    internal class TransitionAdjacencyMultiplierMapEntry
    {
        public Dictionary<double, TransitionAdjacencyMultiplier> Map = new Dictionary<double, TransitionAdjacencyMultiplier>();
        public SortedList<double, TransitionAdjacencyMultiplier> Items = new SortedList<double, TransitionAdjacencyMultiplier>();
    }

    internal class TransitionAdjacencyMultiplierMap : STSimMapBase4<TransitionAdjacencyMultiplierMapEntry>
    {
        private STSimDistributionProvider m_DistributionProvider;

        public TransitionAdjacencyMultiplierMap(
            Scenario scenario, 
            TransitionAdjacencyMultiplierCollection multipliers,
            STSimDistributionProvider distributionProvider) : base(scenario)
        {

            this.m_DistributionProvider = distributionProvider;

            foreach (TransitionAdjacencyMultiplier Item in multipliers)
            {
                this.AddMultiplier(Item);
            }
        }

        public double GetAdjacencyMultiplier(
            int transitionGroupId, int stratumId, int? secondaryStratumId, 
            int? tertiaryStratumId, int iteration, int timestep, double attributeValue)
        {
            TransitionAdjacencyMultiplierMapEntry e = this.GetItem(
                transitionGroupId, stratumId, secondaryStratumId, tertiaryStratumId, iteration, timestep);

            if (e == null)
            {
                return 1.0;
            }

            Debug.Assert(e.Items.Count == e.Map.Count);

            if (e.Map.Count == 1)
            {
                TransitionAdjacencyMultiplier tam = e.Items.First().Value;
                tam.Sample(iteration, timestep, this.m_DistributionProvider, DistributionFrequency.Always);

                return tam.CurrentValue.Value;
            }

            if (e.Map.ContainsKey(attributeValue))
            {
                TransitionAdjacencyMultiplier tam = e.Map[attributeValue];
                tam.Sample(iteration, timestep, this.m_DistributionProvider, DistributionFrequency.Always);

                return tam.CurrentValue.Value;
            }

            double PrevVal = -1.0;
            double NextVal = -1.0;

            Debug.Assert(attributeValue >= 0.0);

            foreach (double k in e.Items.Keys)
            {
                Debug.Assert(k != attributeValue);

                if (k > attributeValue)
                {
                    NextVal = k;
                    break;
                }

                PrevVal = k;
            }

            if (PrevVal == -1.0)
            {
                PrevVal = e.Items.First().Key;
            }

            if (NextVal == -1.0)
            {
                NextVal = PrevVal;
            }

            TransitionAdjacencyMultiplier PrevMult = e.Map[PrevVal];
            TransitionAdjacencyMultiplier NextMult = e.Map[NextVal];

            PrevMult.Sample(iteration, timestep, this.m_DistributionProvider, DistributionFrequency.Always);
            NextMult.Sample(iteration, timestep, this.m_DistributionProvider, DistributionFrequency.Always);

            if (PrevVal != -1.0)
            {
                if (NextVal != -1.0)
                {
                    if (PrevVal == NextVal)
                    {
                        return PrevMult.CurrentValue.Value;
                    }
                    else
                    {
                        return MathUtils.Interpolate(
                            PrevVal, 
                            PrevMult.CurrentValue.Value, 
                            NextVal, 
                            NextMult.CurrentValue.Value, 
                            attributeValue);
                    }
                }
                else
                {
                    return PrevMult.CurrentValue.Value;
                }
            }
            else
            {
                return 1.0;
            }
        }

        private void AddMultiplier(TransitionAdjacencyMultiplier item)
        {
            TransitionAdjacencyMultiplierMapEntry e = this.GetItemExact(
                item.TransitionGroupId, item.StratumId, item.SecondaryStratumId,
                item.TertiaryStratumId, item.Iteration, item.Timestep);

            if (e == null)
            {
                e = new TransitionAdjacencyMultiplierMapEntry();

                this.AddItem(item.TransitionGroupId, item.StratumId, item.SecondaryStratumId, 
                    item.TertiaryStratumId, item.Iteration, item.Timestep, e);
            }

            if (!e.Map.ContainsKey(item.AttributeValue))
            {
                Debug.Assert(!e.Items.ContainsKey(item.AttributeValue));

                e.Map.Add(item.AttributeValue, item);
                e.Items.Add(item.AttributeValue, item);
            }

            Debug.Assert(this.HasItems);
        }
    }
}
